flink-runtime/src/main/java/org/apache/flink/runtime/rest/FileUploadHandler.java

```diff
@@ -106,7 +106,7 @@ protected void channelRead0(final ChannelHandlerContext ctx, final HttpObject ms
 				final HttpContent httpContent = (HttpContent) msg;
 				currentHttpPostRequestDecoder.offer(httpContent);
 
-				while (currentHttpPostRequestDecoder.hasNext()) {
+				while (httpContent != LastHttpContent.EMPTY_LAST_CONTENT && currentHttpPostRequestDecoder.hasNext()) {
```

Comment:
Why do we have to add this filter condition?

Reply:
because it fails otherwise, see https://issues.apache.org/jira/browse/FLINK-9500. See PR description on how to reproduce it.

Reply:
Why does it fail otherwise?

Reply:
When you have passed an `EMPTY_LAST_HTTP_CONTENT` to the decoder the entire content of the message was already processsed. Calling `hasNext()` on a decoder that has already processed the entire content throws the exception.

This was done to differentiate between "I don't have more data _right now_." and "I will never have more date.". We manually do this by checking `instanceof LastHttpContent`.

Reply:
The `HttpPostRequestEncoder` always sends an `EMPTY_LAST_HTTP_CONTENT` to mark the end of the requests, and the decoder handles it as described.

It could be that the decoder is not actually a general-purpose decoder but explicitly written as a counterpart for the encoder.

Reply:
But we are using the `HttpPostRequestEncoder`, right? So do we use it maybe wrongly?

Reply:
As far as i can tell we are using the encoder correctly, but the decoder usage wasn't written against the encoder, but (i guess) only against `curl` or the web UI that never send an empty LAST_HTTP_CONTENT, but a `DefaultLastHttpContent` instead.

Interestingly enough, if you use anything but the netty encoder, without an `instanceof LastHttpContent` check it isn't possible to know whether the decoder is done or not.

Reply:
@zentol probably I do not fully grasp this part of the code here but why do we compare references here? Isn't that about receiving data that was encoded on a different machine? :thinking: I'm asking because FLINK-29326 popped up. I would expected something like `instanceof LastHttpContent` here :thinking: